### PR TITLE
[SPARKR-221] include grouping columns in agg()

### PR DESCRIPTION
**Pull request by davies**
add docs for groupBy() and agg()

### CODE REVIEW

**Comment by shivaram** [comment]
@cafreeman Could you also take a look at this ?

File: pkg/R/group.R [hunk 1]
```diff
@@ -1,19 +1,39 @@
...
+#' Count
+#'
+#' Count the number of rows for each group.
+#' The resulting DataFrame will also contain the grouping columns.
+#'
+#' @param x a GroupedData
+#' @return a DataFrame
+#' @export
+#' @examples
+#' \dontrun {
```
Comment (cafreeman) [inline pkg/R/group.R]: Is there supposed to be an example here?

Comment (davies) [inline pkg/R/group.R]: yes, missed

File: pkg/src/src/main/scala/edu/berkeley/cs/amplab/sparkr/SQLUtils.scala [hunk 2]
```diff
@@ -18,6 +15,22 @@ object SQLUtils {
     arr.toSeq
   }
 
+  // A helper to include grouping columns in Agg()
```
Comment (shivaram) [inline pkg/src/src/main/scala/edu/berkeley/cs/amplab/sparkr/SQLUtils.scala]: This method works for now. Could you add a TODO here saying we can get rid of this after the merge (and just call the spark-internal API without reflection)

File: pkg/src/src/main/scala/edu/berkeley/cs/amplab/sparkr/SQLUtils.scala [hunk 3]
```diff
@@ -18,6 +15,22 @@ object SQLUtils {
...
+  def aggWithGrouping(gd: GroupedData, exprs: Column*): DataFrame = {
+    val aggExprs = exprs.map{ col =>
```
Comment (shivaram) [inline pkg/src/src/main/scala/edu/berkeley/cs/amplab/sparkr/SQLUtils.scala]: minor style nit: space between `map` and `{`

File: pkg/R/DataFrame.R [hunk 4]
```diff
@@ -697,8 +697,23 @@ setMethod("toRDD",
...
+#' @return a GroupedData
+#' @seealso GroupedData
+#' @rdname DataFrame
+#' @export
+#' @examples
+#' \dontrun {
+#'   # Compute the average for all numeric columns grouped by department.
+#'   avg(groupBy(df, "department"))
+#'
+#'   // Compute the max age and average salary, grouped by department and gender.
```
Comment (shivaram) [inline pkg/R/DataFrame.R]: Replace `//` with `#` here.

**Comment by shivaram** [comment]
Thanks @davies for fixing this. Some minor comments but otherwise this looks good. 
BTW its great that you could find some Scala method for it but its unfortunate we need to use reflection -- We can try to find a better fix in 1.4

@cafreeman any other comments ?

**Comment by cafreeman** [comment]
I think this looks great. My only question is about how we document `GroupedData`. This is totally a nitpick, but is calling the class "a GroupedData" in the documentation correct? Is it not like... a GroupedDataFrame or something?

Beyond that, this is looking good and the Scala method is a solid fix for the groupBy columns in the aggregation function.

**Comment by shivaram** [comment]
Hmm I think the name is mostly to reflect the name chosen on the [Scala side]( https://github.com/apache/spark/blob/3980ebdf188d77799b55b407b115cdc82f51d532/sql/core/src/main/scala/org/apache/spark/sql/GroupedData.scala) -- I see your point and we can probably change how it gets printed to the user.

BTW we should try to add a `print` function for these classes so that when you just enter `df` on the R console after creating a DataFrame it shows something like `DataFrame with columns age:Int, name:String` etc.

**Comment by cafreeman** [comment]
Can you overload `print` for S4 classes?  Does it work the same way as with S3? (e.g. will the standard `print.DataFrame` syntax work?

**Comment by shivaram** [comment]
I think you can write a `setMethod("print")` in the class and it should work

**Comment by davies** [comment]
@shivaram It turns out should be `show` for S3 class, added.

**Comment by davies** [comment]
sbt.ResolveException: download failed: org.apache.spark#spark-sql_2.10;1.3.0-rc2!spark-sql_2.10.jar
	at sbt.IvyActions$.sbt$IvyActions$$resolve(IvyActions.scala:243)
	at sbt.IvyActions$$anonfun$updateEither$1.apply(IvyActions.scala:158)
	at sbt.IvyActions$$anonfun$updateEither$1.apply(IvyActions.scala:156)
	at sbt.IvySbt$Module$$anonfun$withModule$1.apply(Ivy.scala:147)
	at sbt.IvySbt$Module$$anonfun$withModule$1.apply(Ivy.scala:147)
	at sbt.IvySbt$$anonfun$withIvy$1.apply(Ivy.scala:124)
	at sbt.IvySbt.sbt$IvySbt$$action$1(Ivy.scala:56)
	at sbt.IvySbt$$anon$3.call(Ivy.scala:64)
	at xsbt.boot.Locks$GlobalLock.withChannel$1(Locks.scala:93)
	at xsbt.boot.Locks$GlobalLock.xsbt$boot$Locks$GlobalLock$$withChannelRetries$1(Locks.scala:78)
	at xsbt.boot.Locks$GlobalLock$$anonfun$withFileLock$1.apply(Locks.scala:97)
	at xsbt.boot.Using$.withResource(Using.scala:10)
	at xsbt.boot.Using$.apply(Using.scala:9)
	at xsbt.boot.Locks$GlobalLock.ignoringDeadlockAvoided(Locks.scala:58)
	at xsbt.boot.Locks$GlobalLock.withLock(Locks.scala:48)
	at xsbt.boot.Locks$.apply0(Locks.scala:31)
	at xsbt.boot.Locks$.apply(Locks.scala:28)
	at sbt.IvySbt.withDefaultLogger(Ivy.scala:64)
	at sbt.IvySbt.withIvy(Ivy.scala:119)
	at sbt.IvySbt.withIvy(Ivy.scala:116)
	at sbt.IvySbt$Module.withModule(Ivy.scala:147)
	at sbt.IvyActions$.updateEither(IvyActions.scala:156)
	at sbt.Classpaths$$anonfun$sbt$Classpaths$$work$1$1.apply(Defaults.scala:1282)
	at sbt.Classpaths$$anonfun$sbt$Classpaths$$work$1$1.apply(Defaults.scala:1279)
	at sbt.Classpaths$$anonfun$doWork$1$1$$anonfun$84.apply(Defaults.scala:1309)
	at sbt.Classpaths$$anonfun$doWork$1$1$$anonfun$84.apply(Defaults.scala:1307)
	at sbt.Tracked$$anonfun$lastOutput$1.apply(Tracked.scala:35)
	at sbt.Classpaths$$anonfun$doWork$1$1.apply(Defaults.scala:1312)
	at sbt.Classpaths$$anonfun$doWork$1$1.apply(Defaults.scala:1306)
	at sbt.Tracked$$anonfun$inputChanged$1.apply(Tracked.scala:45)
	at sbt.Classpaths$.cachedUpdate(Defaults.scala:1324)
	at sbt.Classpaths$$anonfun$updateTask$1.apply(Defaults.scala:1264)
	at sbt.Classpaths$$anonfun$updateTask$1.apply(Defaults.scala:1242)
	at scala.Function1$$anonfun$compose$1.apply(Function1.scala:47)
	at sbt.$tilde$greater$$anonfun$$u2219$1.apply(TypeFunctions.scala:40)
	at sbt.std.Transform$$anon$4.work(System.scala:63)
	at sbt.Execute$$anonfun$submit$1$$anonfun$apply$1.apply(Execute.scala:226)
	at sbt.Execute$$anonfun$submit$1$$anonfun$apply$1.apply(Execute.scala:226)
	at sbt.ErrorHandling$.wideConvert(ErrorHandling.scala:17)
	at sbt.Execute.work(Execute.scala:235)
	at sbt.Execute$$anonfun$submit$1.apply(Execute.scala:226)
	at sbt.Execute$$anonfun$submit$1.apply(Execute.scala:226)
	at sbt.ConcurrentRestrictions$$anon$4$$anonfun$1.apply(ConcurrentRestrictions.scala:159)
	at sbt.CompletionService$$anon$2.call(CompletionService.scala:28)
	at java.util.concurrent.FutureTask.run(FutureTask.java:262)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:471)
	at java.util.concurrent.FutureTask.run(FutureTask.java:262)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
	at java.lang.Thread.run(Thread.java:745)
[error] (*:update) sbt.ResolveException: download failed: org.apache.spark#spark-sql_2.10;1.3.0-rc2!spark-sql_2.10.jar
[error] Total time: 275 s, completed Mar 13, 2015 7:01:32 AM

**Comment by shivaram** [comment]
I restarted the build

File: pkg/src/src/main/scala/edu/berkeley/cs/amplab/sparkr/SQLUtils.scala (hunk 2 above)
Comment (davies) [inline pkg/src/src/main/scala/edu/berkeley/cs/amplab/sparkr/SQLUtils.scala]: done

File: pkg/R/DataFrame.R [hunk 5]
```diff
@@ -171,6 +171,15 @@ setMethod("showDF",
             cat(callJMethod(x@sdf, "showString", numToInt(numRows)))
           })
 
+setMethod("show", "DataFrame",
+          function(object) {
+            cols <- lapply(dtypes(object), function(l) {
+              paste(l, collapse = ":")
+            })
+            s <- paste(cols, collapse = ", ")
+            cat(paste("DataFrame[", s, "]\n"))
```
Comment (shivaram) [inline pkg/R/DataFrame.R]: minor nit: you probably want `sep=""` here. Right now it looks like 
```
DataFrame[ cust_id:string, date:string, day_num:bigint, extended_price:double, sku:bigint, unit_price:double, units:bigint ]
```
so there is an extra space at the beginning and end

**Comment by shivaram** [comment]
Thanks -- LGTM. Merging this

**Pull request merged**